anvil/src/eth/otterscan.rs

```diff
@@ -0,0 +1,202 @@
+        last_page: bool,
+    ) -> Result<Self> {
+        let txs: Vec<Transaction> = join_all(
+            hashes
+                .iter()
+                .map(|hash| async { backend.transaction_by_hash(*hash).await.unwrap().unwrap() }),
+        )
+        .await;
+
+        let receipts: Vec<OtsTransactionReceipt> = join_all(hashes.iter().map(|hash| async {
+            let receipt = backend.transaction_receipt(*hash).await.unwrap().unwrap();
+            let timestamp =
+                backend.get_block(receipt.block_number.unwrap()).unwrap().header.timestamp;
+            OtsTransactionReceipt { receipt, timestamp }
+        }))
+        .await;
+
+        Ok(Self { txs, receipts, first_page, last_page })
+    }
+}
+
+impl OtsInternalOperation {
+    pub fn batch_build(traces: Vec<Trace>) -> Vec<OtsInternalOperation> {
+        traces
+            .iter()
+            .map(|trace| {
+                match (trace.action.clone(), trace.result.clone()) {
+                    (Action::Call(Call { from, to, value, .. }), _) if !value.is_zero() => {
+                        Some(Self { r#type: OtsInternalOperationType::Transfer, from, to, value })
+                    }
+                    (
+                        Action::Create(Create { from, value, .. }),
+                        Some(Res::Create(CreateResult { address, .. })),
+                    ) => Some(Self {
+                        r#type: OtsInternalOperationType::Create,
+                        from,
+                        to: address,
+                        value,
+                    }),
+                    (Action::Suicide(Suicide { address, .. }), _) => {
+                        // can we correctly assume that any suicide has a parent trace?
+                        let (from, value) =
+                            Self::find_suicide_caller(&traces, &trace.trace_address).unwrap();
+
+                        Some(Self {
+                            r#type: OtsInternalOperationType::SelfDestruct,
+                            from,
+                            to: address,
+                            value,
+                        })
+                    }
+                    _ => None,
+                }
+            })
+            .flatten()
+            .collect()
+    }
+
+    // finds the trace that parents a given trace_address
+    fn find_suicide_caller(
+        traces: &Vec<Trace>,
+        suicide_address: &Vec<usize>,
+    ) -> Option<(Address, U256)> {
+        traces
+            .iter()
+            .find(|t| t.trace_address == &suicide_address[..suicide_address.len() - 1])
+            .map(|t| match t.action {
+                Action::Call(Call { from, value, .. }) => (from, value),
+
+                Action::Create(Create { from, value, .. }) => (from, value),
+
+                // TODO can a suicide trace be parented by another suicide?
+                Action::Suicide(_) => Self::find_suicide_caller(traces, &t.trace_address).unwrap(),
```

Comment:
i'm not sure about this? @mattsse you've probably seen more cursed SELFDESTRUCT stuff than me due to reth